Zephyr utils

mahdiehmalekian opened the pull request:
`coordinate_systems.py ` : Contains namedtuples for Zephyr and Cartesian coordinates and the conversion functions between the two
`plane_shift.py` : Contains the class for displacement of nodes in the Cartesian plane.
`node_edge.py`: Contains the class for edge, edge of a Zephyr graph (which checks whether the edge exists in a perfect yield Zephyr graph), and the node of a Zephyr graph
`qfloor.py`: Contains the class for working with a tile (subset of nodes) in quotient Zephyr. Also contains the class for putting the tiles together in a number of rows and columns
`survey.py`: Contains the class for taking a survey of a Zephyr graph compared to a perfect-yield Zephyr graph on the same grid and tile size
The test suite contains the tests for these modules.

jackraymond committed:
Improve defaulting of max_num_emb; move basic helper function from feasibility to parallel_embeddings

jackraymond committed:
Correct lattice_size signature per review request

JoelPasvolsky committed:
Restructure docs (minimal changes)

JoelPasvolsky committed:
Update header and conf.py fixes

JoelPasvolsky committed:
Add Usage Info section, fix api_ref anchor

JoelPasvolsky committed:
Fix SDK build warning for repo (minimal until last rebase)

jackraymond committed:
Simplify tiling branch behaviour (single fixed embedding). Add timeout parameter

Commit:
Initial commit

Commit:
Reformat

Commit:
Reformat

JoelPasvolsky committed:
Update theme and xrefs

JoelPasvolsky committed:
Flatten ref structure

JoelPasvolsky committed:
Fix missing C++ functions

JoelPasvolsky committed:
Remove the ``find_embedding:189: ERROR: Unexpected indentation``  from multiple package builds

JoelPasvolsky committed:
Update missed path to readme

JoelPasvolsky committed:
Turn off dot (graphviz) to prevent multiple build warnings

JoelPasvolsky committed:
Remove leftover todo (thanks @thisac)

JoelPasvolsky committed:
Remove obsolete doxygen configuration

Commit:
Allow Edge input to be any type, not just int + Add check_edge_valid flag to ZEdge

Commit:
Turn off check_edge_valid flag for _edges attribute in ZSurvey

jackraymond committed:
Add test of timeout

jackraymond committed:
Fiddle with test_parallel_embeddings to satisfy arbitrary circleCI complaints

jackraymond committed:
Make dwave.embedding optional dependency

jackraymond committed:
basic version of imported function is_valid_embedding

jackraymond committed:
Correct variable names in fallover function

jackraymond committed:
random review: Update minorminer/utils/parallel_embeddings.py

Co-authored-by: Radomir Stevanovic <radomir.stevanovic@gmail.com>

jackraymond committed:
Apply suggestions from code review

randomir review implementation

Co-authored-by: Radomir Stevanovic <radomir.stevanovic@gmail.com>

jackraymond committed:
Implement randomir review recommendations

jackraymond committed:
Add test of array_to_embeddings

jackraymond committed:
Delete redundant failover code

jackraymond committed:
Apply suggestions from code review

thisac review code suggestions

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

jackraymond committed:
implement thisac review

randomir committed:
Release 0.2.18

thisac commented on minorminer/utils/zephyr/__init__.py:
```diff
@@ -0,0 +1,17 @@
+# Copyright 2025 D-Wave Systems Inc.
```
```suggestion
# Copyright 2025 D-Wave
```
Same in other new files (don't need to change the old ones).

mahdiehmalekian replied:
Done in commit e6eca6d

thisac replied:
Thanks, although I think you missed this file (the `__init__.py`).

mahdiehmalekian replied:
Thanks. Done in commit [3365f07](https://github.com/dwavesystems/minorminer/pull/261/commits/3365f07a9f7b912d76bfbe768054fcae62241369)

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+        self._edge = self._set_edge(x, y)
+
+    def _set_edge(self, x, y):
+        if x < y:
```
Missing docstring. I think a single line one is useful even for private methods and functions.

mahdiehmalekian replied:
Done in commit [20b79d0]((https://github.com/dwavesystems/minorminer/pull/261/commits/20b79d020fb5170cad6ba883d71791ce63231752))

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+
+
+class ZEdge(Edge):
+    """Initializes a ZEdge with 'ZNode' nodes x, y
```
```suggestion
    """Initializes a ZEdge with 'ZNode' nodes x, y.
```

mahdiehmalekian replied:
Done in commit https://github.com/dwavesystems/minorminer/commit/20b79d020fb5170cad6ba883d71791ce63231752

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+        x (ZNode): One endpoint of edge.
+        y (ZNode): Another endpoint of edge.
+        check_edge_valid (bool, optional): Flag to whether check the validity of values and types of x, y.
+        Defaults to True.
```
```suggestion
        check_edge_valid (bool, optional): Flag to whether check the validity of values and types of ``x``, ``y``.
            Defaults to True.
```

mahdiehmalekian replied:
Done in commit https://github.com/dwavesystems/minorminer/commit/20b79d020fb5170cad6ba883d71791ce63231752

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+        TypeError: If either of x or y is not 'ZNode'.
+        ValueError: If x, y do not have the same shape.
+        ValueError: If x, y are not neighbours in a perfect yield (quotient)
+        Zephyr graph.
```
```suggestion
        ValueError: If x, y are not neighbours in a perfect yield (quotient)
            Zephyr graph.
```

mahdiehmalekian replied:
Done in commit https://github.com/dwavesystems/minorminer/commit/20b79d020fb5170cad6ba883d71791ce63231752

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+        y: ZNode,
+        check_edge_valid: bool = True,
+    ) -> None:
+        if check_edge_valid:
+            if not isinstance(x, ZNode) or not isinstance(y, ZNode):
+                raise TypeError(f"Expected x, y to be ZNode, got {type(x), type(y)}")
+            if x.shape != y.shape:
+                raise ValueError(f"Expected x, y to have the same shape, got {x.shape, y.shape}")
```
```suggestion
            if not isinstance(x, ZNode) or not isinstance(y, ZNode):
                raise TypeError(f"Expected x, y to be ZNode, got {type(x), type(y)}")

            if x.shape != y.shape:
                raise ValueError(f"Expected x, y to have the same shape, got {x.shape, y.shape}")

```

mahdiehmalekian replied:
Done in commit [9f7e5ab](https://github.com/dwavesystems/minorminer/pull/261/commits/9f7e5ab841c21b74bb7ddb875472aa38faf93a14)

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+
+    Args:
+        x (ZNode): One endpoint of edge.
+        y (ZNode): Another endpoint of edge.
```
```suggestion
        x (ZNode): Endpoint of edge. Must have same shape as ``y``.
        y (ZNode): Endpoint of edge. Must have same shape as ``x``
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+                    kind_found = True
+                    break
+            if not kind_found:
+                raise ValueError(f"Expected x, y to be neighbours, got {x, y}")
```
Could be simplified to:
```suggestion
            for kind in ("internal", "external", "odd"):
                if x.is_neighbor(y, nbr_kind=kind):
                    break
            else:
                raise ValueError(f"Expected x, y to be neighbours, got {x, y}")
```

mahdiehmalekian replied:
[9f7e5ab](https://github.com/dwavesystems/minorminer/pull/261/commits/9f7e5ab841c21b74bb7ddb875472aa38faf93a14)

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+        m: grid size, t: tile size
+        Defaults to None.
+        convert_to_z (bool | None, optional): Whether to express the coordinates in ZephyrCoordinates.
+        Defaults to None.
```
```suggestion
        convert_to_z (bool | None, optional): Whether to express the coordinates in ZephyrCoordinates.
            Defaults to None.
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+    """Initializes 'ZNode' with coord and optional shape.
+
+    Args:
+        coord (CartesianCoord | ZephyrCoord | tuple[int]): coordinate in (quotient) Zephyr or (quotient) Cartesian
```
```suggestion
        coord (CartesianCoord | ZephyrCoord | tuple[int]): Coordinate in (quotient) Zephyr or (quotient) Cartesian
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+
+    Args:
+        coord (CartesianCoord | ZephyrCoord | tuple[int]): coordinate in (quotient) Zephyr or (quotient) Cartesian
+        shape (ZShape | tuple[int | None] | None, optional): shape of Zephyr graph containing ZNode.
+        m: grid size, t: tile size
+        Defaults to None.
```
Maybe clarify this argument a bit. I think the mentions of `m` and `t` here are a bit lost, and `ZShape` could be mentioned as a namedtuple containing grid and tile size.

mahdiehmalekian replied:
Done in commit [974489c](https://github.com/dwavesystems/minorminer/pull/261/commits/974489cc015e2990ee4df22953be9a4e374857f7)

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+
+    Note: If the given coord has non-None k value (in either Cartesian or Zephyr coordinates),
+        shape = None raises ValueError. In this case the tile size of Zephyr, t,
+        must be provided.
```
```suggestion
    Note: 
        If the given ``coord`` has non-None k value (in either Cartesian or Zephyr coordinates),
        ``shape = None`` raises ``ValueError``. In this case the tile size of Zephyr, t,
        must be provided.
```

thisac replied:
Also, it's not clear here what `k` and `t` are.

mahdiehmalekian replied:
Done in commit [974489c](https://github.com/dwavesystems/minorminer/pull/261/commits/974489cc015e2990ee4df22953be9a4e374857f7)

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+    def shape(self, new_shape) -> None:
+        """Sets a new value for shape"""
+        if new_shape is None:
+            new_shape = ZShape()
```
I think this is better to have in the init.
```suggestion
```
like this
```python
if shape:
    self.shape = shape
else:
    shape = ZShape()
```

mahdiehmalekian replied:
Done in commit 974489c

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+        for var, val in {"m": new_shape.m, "t": new_shape.t}.items():
+            if (val is not None) and (not isinstance(val, int)):
+                raise TypeError(f"Expected {var} to be None or 'int', got {type(val)}")
+        self._shape = new_shape
```
Empty lines make things more readable here IMO.
```suggestion

        if not isinstance(new_shape, ZShape):
            raise TypeError(
                f"Expected shape to be tuple[int | None] or ZShape or None, got {type(new_shape)}"
            )

        if hasattr(self, "_ccoord"):
            if (self._ccoord.k is None) != (new_shape.t is None):
                raise ValueError(
                    f"ccoord, shape must be both quotient or non-quotient, got {self._ccoord, new_shape}"
                )

        for var, val in {"m": new_shape.m, "t": new_shape.t}.items():
            if (val is not None) and (not isinstance(val, int)):
                raise TypeError(f"Expected {var} to be None or 'int', got {type(val)}")

        self._shape = new_shape
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+                f"Expected shape to be tuple[int | None] or ZShape or None, got {type(new_shape)}"
+            )
+        if hasattr(self, "_ccoord"):
+            if (self._ccoord.k is None) != (new_shape.t is None):
```
`self._ccoord` should maybe be set in the init (at least to `None`, but it seems like that might have another meaning in this class). This way it's easier to keep track of the classes available attributes, especially since there's a `ccoord` property below that returns `self._ccoord`.

In that case the `hasattr` check here isn't needed anymore.

mahdiehmalekian replied:
I need want self._ccoord  to have `k` field. So, if I simply set it to None in init, I need to add other code, which I think makes it ultimately noisier.

thisac replied:
You could instead check for `self._ccoord.k` with `getattr(self._ccoord, "k", None)`, which returns `None` if `self._ccoord` is `None`.

mahdiehmalekian replied:
Dealt with in a56aa8453ea0915c9c4b824e5606939d762a3561

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+                    f"ccoord, shape must be both quotient or non-quotient, got {self._ccoord, new_shape}"
+                )
+        for var, val in {"m": new_shape.m, "t": new_shape.t}.items():
+            if (val is not None) and (not isinstance(val, int)):
```
You can access the item names using `new_shape._fields` (giving `('m', 't')`) or transforming it to a dictionary directly using `new_shape._asdict`, instead of having to create a new dict with the fields hardcoded here again.

mahdiehmalekian replied:
Done in commit https://github.com/dwavesystems/minorminer/pull/261/commits/974489cc015e2990ee4df22953be9a4e374857f7

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+            raise TypeError(
+                f"Expected ccoord to be CartesianCoord or tuple[int], got {type(new_ccoord)}"
+            )
+        for c in (new_ccoord.x, new_ccoord.y):
```
```suggestion
        for c in new_ccoord:
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+                f"Expected ccoord.x and ccoord.y to differ in parity, got {new_ccoord.x, new_ccoord.y}"
+            )
+        # check k value of CartesianCoord is consistent with t
+        if hasattr(self, "_shape"):
```
Again, `self._shape` should probably be set to `None` in init, and this check omitted.

mahdiehmalekian replied:
I need want self._shape to have `m` field and `t` field. So, if I simply set it to None in init, I need to add other code, which I think makes it ultimately noisier.

mahdiehmalekian replied:
Dealt with in a56aa8453ea0915c9c4b824e5606939d762a3561

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+    @staticmethod
+    def get_coord(coord: tuple[int]) -> CartesianCoord | ZephyrCoord:
+        """Takes a tuple[int] and returns the corresponding CartesianCoord
+        or ZephyrCoord"""
```
```suggestion
        """Takes a tuple[int] and returns the corresponding ``CartesianCoord`` or ``ZephyrCoord``"""
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+        self._ccoord = new_ccoord
+
+    @staticmethod
+    def get_coord(coord: tuple[int]) -> CartesianCoord | ZephyrCoord:
```
Would it make sense renaming this `tuple_to_coord` or `create_coord` instead? The former probably makes sense since you also have `cartesian_to_zephyr` below.

mahdiehmalekian replied:
Done in https://github.com/dwavesystems/minorminer/pull/261/commits/70d84ed616e04331a0f181015fdf1fb3e2774596

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+                ZephyrCoord(u=u, w=w, j=j, z=z)
+                if len(k) == 0
+                else ZephyrCoord(u=u, w=w, k=k[0], j=j, z=z)
+            )
```
I usually prefer _not_ using the `... if ... else ...` and instead write it out:
```suggestion
            if len(k) == 0
                return ZephyrCoord(u=u, w=w, j=j, z=z)
            else 
                return ZephyrCoord(u=u, w=w, k=k[0], j=j, z=z)
```

mahdiehmalekian replied:
Done in https://github.com/dwavesystems/minorminer/pull/261/commits/70d84ed616e04331a0f181015fdf1fb3e2774596

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+            return CartesianCoord(x=x, y=y) if len(k) == 0 else CartesianCoord(x=x, y=y, k=k[0])
+        if len_coord in (4, 5):
+            u, w, *k, j, z = coord
+            for var, val in {"u": u, "j": j}.items():
```
Cleaner to just use tuples.
```suggestion
            for var, val in [("u", u), ("j", j)]:
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+        return self._ccoord.x % 2
+
+    def is_vertical(self) -> bool:
+        """Decides whether self is a vertical qubit"""
```
`self` is a node, right?
```suggestion
        """Decides whether node is a vertical qubit"""
```
Similarly for docstrings below.

mahdiehmalekian replied:
Done in https://github.com/dwavesystems/minorminer/pull/261/commits/70d84ed616e04331a0f181015fdf1fb3e2774596

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+
+    @property
+    def direction(self) -> int:
+        """Returns direction, 0 or 1"""
```
What is meant by the direction of of a node? 
```suggestion
        """Returns direction, 0 (vertical) or 1 (horizontal)."""
```

mahdiehmalekian replied:
Done in https://github.com/dwavesystems/minorminer/pull/261/commits/70d84ed616e04331a0f181015fdf1fb3e2774596

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+        if not isinstance(other, ZNode):
+            other = ZNode(other)
+        if self._shape != other._shape:
+            return
```
Conventions are to return `None` explicitly if function returns anything. See https://stackoverflow.com/a/15300671.
```suggestion
            return None
```
Same in other places in this method.

mahdiehmalekian replied:
Done in commit [fd78154](https://github.com/dwavesystems/minorminer/pull/261/commits/fd78154bb8a05b02fcd4f8cc75569591166c3272)

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+        other: ZNode,
+    ) -> str | None:
+        """Returns the kind of coupler between self and other,
+        'internal', 'external', 'odd', or None."""
```
This, and a couple of more methods in this class, are missing complete docstrings (with Args, Returns, etc.)

mahdiehmalekian replied:
Done in commit https://github.com/dwavesystems/minorminer/commit/20b79d020fb5170cad6ba883d71791ce63231752

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+        other: ZNode,
+    ) -> str | None:
+        """Returns the kind of coupler between self and other,
+        'internal', 'external', 'odd', or None."""
```
Instead of returning string representations of the coupler kind, consider using an [enum](https://docs.python.org/3/howto/enum.html) instead:

mahdiehmalekian replied:
Done in commit https://github.com/dwavesystems/minorminer/commit/20b79d020fb5170cad6ba883d71791ce63231752

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+            if y1 != y2:  # odd, external neighbors only on the same horizontal lines
+                return
+            diff_x = abs(x1 - x2)
+            return "odd" if diff_x == 2 else "external" if diff_x == 4 else None
```
Nit-picky here, but I'd either skip the final else here and add a comment,
```suggestion

        # else, if horizontal
        if y1 != y2:  # odd, external neighbors only on the same horizontal lines
            return
        diff_x = abs(x1 - x2)
        return "odd" if diff_x == 2 else "external" if diff_x == 4 else None
```
or add a final `return None` to clarify that no neighbours found
```suggestion
        else:
            if y1 != y2:  # odd, external neighbors only on the same horizontal lines
                return
            diff_x = abs(x1 - x2)
            return "odd" if diff_x == 2 else "external" if diff_x == 4 else None

    # no neighbors found
    return None
```

mahdiehmalekian replied:
Done in commit https://github.com/dwavesystems/minorminer/commit/20b79d020fb5170cad6ba883d71791ce63231752

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+        self,
+        where: Callable[[CartesianCoord | ZephyrCoord], bool] = lambda coord: True,
+    ) -> Generator[ZNode]:
+        """Generator of internal neighbors"""
```
Missing details in docstring.

mahdiehmalekian replied:
Done in commit https://github.com/dwavesystems/minorminer/commit/20b79d020fb5170cad6ba883d71791ce63231752

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,635 @@
+            except GeneratorExit:
+                raise
+            except Exception:
+                pass
```
Never good to except a generic `Exception`. Better to be explicit and/or do checks prior to yielding `ZNode`.

mahdiehmalekian replied:
Done in commit https://github.com/dwavesystems/minorminer/commit/20b79d020fb5170cad6ba883d71791ce63231752

thisac reviewed (commented):
Thanks for the hard work @mahdiehmalekian. Had a first pass through `node_edge`. Will come back to it again, and the other files, in a future review soon.

Commit:
Correct header

Commit:
Add NodeKind and EdgeKind and adjust; Add document

Commit:
Reformat

Commit:
Explicitly return None

Commit:
Simplify, Add space

mahdiehmalekian committed:
Clarify arguments

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Capitalize

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

Commit:
Clarify documentation, set shape in init

Commit:
Add doc to EdgeKind, NodeKind, clarify doc re. direction, Change name

Commit:
Edit document

Commit:
Change nbr_kind argument type

Commit:
Add edge_kind to ZEdge

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+
+    Args:
+        x : One endpoint of edge.
+        y : Another endpoint of edge.
```
```suggestion
        x: One endpoint of edge.
        y: Another endpoint of edge.
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+        self,
+        x,
+        y,
+    ) -> None:
```
```suggestion
    def __init__(self, x: int , y: int) -> None:
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+    ) -> None:
+        self._edge = self._set_edge(x, y)
+
+    def _set_edge(self, x, y):
```
```suggestion
    def _set_edge(self, x: int, y: int) -> tuple[int, int]:
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+    def __getitem__(self, index: int) -> int:
+        return self._edge[index]
+
+    def __eq__(self, other: Edge):
```
```suggestion
    def __eq__(self, other: Edge) -> bool:
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+
+    @property
+    def edge_kind(self) -> EdgeKind:
+        return self._edge_kind
```
Docstring?

mahdiehmalekian replied:
Done in 1a8b937d5dce9308e495fadcafb3184c96bf7879

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+        coord (CartesianCoord | ZephyrCoord | tuple[int]): Coordinate in (quotient) Zephyr or (quotient) Cartesian
+        shape (ZShape | tuple[int | None] | None, optional): Shape of the Zephyr graph containing this ZNode.
+            If a ZShape is passed, it should be a namedtuple with fields `m` (grid size of the Zephyr graph) and `t` (tile size of the Zephyr graph).
+        Defaults to None.
```
```suggestion
        coord (CartesianCoord | ZephyrCoord | tuple[int]): Coordinate in (quotient) Zephyr or (quotient) Cartesian.
        shape (ZShape | tuple[int | None] | None, optional): Shape of the Zephyr graph containing this ZNode.
            If a ZShape is passed, it should be a namedtuple with fields `m` (grid size of the Zephyr graph) 
            and `t` (tile size of the Zephyr graph). Defaults to None.
```
Just a bit of docstring reformatting.

Commit:
Disable changing znodes of QuoTile, change Dim to Dim_2D

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+        convert_to_z: bool | None = None,
+    ) -> None:
+        if shape:
+            self.shape = shape
```
Since `ZNode.shape` is expected to always return a `ZShape`.
```suggestion
            self.shape = ZShape(*shape)
```

jackraymond committed:
Bugfix: replace full graph by local decimated graph for use tiling branch with testing derived from TilingComposite

jackraymond committed:
Minor changes to docstrings and defaulting for compatibility/comprehensibility

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+                else:
+                    raise ValueError(
+                        f"Expected ccoord.x and ccoord.y to be in {range(4*self._shape.m+1)}, got {new_ccoord.x, new_ccoord.y}"
+                    )
+        self._ccoord = new_ccoord
+
+    @staticmethod
+    def tuple_to_coord(coord: tuple[int]) -> CartesianCoord | ZephyrCoord:
+        """Takes a tuple[int] and returns the corresponding ``CartesianCoord`` or ``ZephyrCoord``"""
+        if (not isinstance(coord, tuple)) or (not all(isinstance(c, int) for c in coord)):
+            raise TypeError(f"Expected {coord} to be a tuple[int], got {coord}")
+        if any(c < 0 for c in coord):
+            raise ValueError(f"Expected elements of coord to be non-negative, got {coord}")
+        len_coord = len(coord)
+        if len_coord in (2, 3):
+            x, y, *k = coord
+            if x % 2 == y % 2:
+                raise ValueError(f"Expected x, y to differ in parity, got {x, y}")
+            return CartesianCoord(x=x, y=y) if len(k) == 0 else CartesianCoord(x=x, y=y, k=k[0])
+        if len_coord in (4, 5):
+            u, w, *k, j, z = coord
+            for var, val in [("u", u), ("j", j)]:
+                if not val in [0, 1]:
+                    raise ValueError(f"Expected {var} to be in [0, 1], got {val}")
+            if len(k) == 0:
+                return ZephyrCoord(u=u, w=w, j=j, z=z)    
+            return ZephyrCoord(u=u, w=w, k=k[0], j=j, z=z)
```
To make this a bit more legible, I'd recommend splitting it into sections. An empty line after raises, and around returns make things easier to parse at a glance, along with around ifs and fors.
```suggestion
        if (not isinstance(coord, tuple)) or (not all(isinstance(c, int) for c in coord)):
            raise TypeError(f"Expected {coord} to be a tuple[int], got {coord}")

        if any(c < 0 for c in coord):
            raise ValueError(f"Expected elements of coord to be non-negative, got {coord}")

        len_coord = len(coord)
        if len_coord in (2, 3):
            x, y, *k = coord
            if x % 2 == y % 2:
                raise ValueError(f"Expected x, y to differ in parity, got {x, y}")

            return CartesianCoord(x=x, y=y) if len(k) == 0 else CartesianCoord(x=x, y=y, k=k[0])

        if len_coord in (4, 5):
            u, w, *k, j, z = coord
            for var, val in [("u", u), ("j", j)]:
                if not val in [0, 1]:
                    raise ValueError(f"Expected {var} to be in [0, 1], got {val}")

            if len(k) == 0:
                return ZephyrCoord(u=u, w=w, j=j, z=z)

            return ZephyrCoord(u=u, w=w, k=k[0], j=j, z=z)
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+        other: ZNode,
+    ) -> EdgeKind | None:
+        """Returns the kind of edge between self and other;
+        EdgeKind if there is an edge in perfect Zephyr between them or None.
```
```suggestion
        """Returns the kind of edge between two ZNodes.

        :class:`EdgeKind` if there is an edge in perfect Zephyr between them or ``None``.
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+        EdgeKind if there is an edge in perfect Zephyr between them or None.
+
+        Args:
+            other (ZNode): a ZNode
```
```suggestion
            other (ZNode): The neigboring :class:`ZNode`.
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+    ) -> EdgeKind | None:
+        """Returns the kind of edge between self and other;
+        EdgeKind if there is an edge in perfect Zephyr between them or None.
+
+        Args:
+            other (ZNode): a ZNode
+
+        Returns:
+            EdgeKind | None: The edge kind between self and other in perfect Zephyr or None if there is no edge between in perfect Zephyr.
+        """
+        if not isinstance(other, ZNode):
+            other = ZNode(other)
+        if self._shape != other._shape:
+            return None
+        coord1 = self._ccoord
+        coord2 = other._ccoord
+        x1, y1 = coord1.x, coord1.y
+        x2, y2 = coord2.x, coord2.y
+        if abs(x1 - x2) == abs(y1 - y2) == 1:
+            return EdgeKind.INTERNAL
+        if x1 % 2 != x2 % 2:
+            return None
+        if coord1.k != coord2.k:  # odd, external neighbors only on the same k
+            return None
+        if self.node_kind is NodeKind.VERTICAL:  # self vertical
+            if x1 != x2:  # odd, external neighbors only on the same vertical lines
+                return None
+            diff_y = abs(y1 - y2)
+            return EdgeKind.ODD if diff_y == 2 else EdgeKind.EXTERNAL if diff_y == 4 else None
+
+        # else, self is horizontal
+        if y1 != y2:  # odd, external neighbors only on the same horizontal lines
+            return None
+        diff_x = abs(x1 - x2)
+        return EdgeKind.ODD if diff_x == 2 else EdgeKind.EXTERNAL if diff_x == 4 else None
```
```suggestion
        if not isinstance(other, ZNode):
            other = ZNode(other)

        if self._shape != other._shape:
            return None

        coord1 = self._ccoord
        coord2 = other._ccoord
        x1, y1 = coord1.x, coord1.y
        x2, y2 = coord2.x, coord2.y
        if abs(x1 - x2) == abs(y1 - y2) == 1:
            return EdgeKind.INTERNAL

        if x1 % 2 != x2 % 2:
            return None

        if coord1.k != coord2.k:  # odd, external neighbors only on the same k
            return None

        if self.node_kind is NodeKind.VERTICAL:  # self vertical
            if x1 != x2:  # odd, external neighbors only on the same vertical lines
                return None
    
            diff_y = abs(y1 - y2)
            return EdgeKind.ODD if diff_y == 2 else EdgeKind.EXTERNAL if diff_y == 4 else None

        # else, self is horizontal
        if y1 != y2:  # odd, external neighbors only on the same horizontal lines
            return None

        diff_x = abs(x1 - x2)
        return EdgeKind.ODD if diff_x == 2 else EdgeKind.EXTERNAL if diff_x == 4 else None
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+                or to `zcoord` if `self.convert_to_z` is True. Defaults to `lambda coord: True`.
+
+        Yields:
+            ZNode: Internal neighbors of self when restricted by `where`.
```
Don't forget that sphinx docstrings should have double back-ticks (``` ``...`` ```) for code words and single back-ticks with roles for classes/methods/attributes, etc. (``:class:`...` ``)
```suggestion
        """Generator of internal neighbors of self when restricted by ``where``.

        Args:
            where (Callable[[CartesianCoord | ZephyrCoord, bool]]):
                A coordinate filter. Applies to ``ccoord`` if ``self.convert_to_z`` is ``False``,
                or to ``zcoord`` if :py:attr:`self.convert_to_z` is ``True``. Defaults to always 

        Yields:
            ZNode: Internal neighbors of self when restricted by `where`.
```

mahdiehmalekian replied:
Partially done in cc2e009a0484d628a96e0ef32748dbbb4c286fbb

mahdiehmalekian replied:
Done in a56aa8453ea0915c9c4b824e5606939d762a3561

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+
+    def internal_neighbors_generator(
+        self,
+        where: Callable[[CartesianCoord | ZephyrCoord], bool] = lambda coord: True,
```
Where might `where` be called with a bool? Seems like it should always take in a coord.

mahdiehmalekian replied:
`where` is a function that takes in coord and outputs bool.

thisac replied:
Oh, yes, sorry. I misread the type hint.

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+            except GeneratorExit:
+                raise
+            except (TypeError, ValueError):
+                pass
```
I don't think you necessarily need this try-except block. You shouldn't need to catch a `GeneratorExit` in this case, and the type error should be handled earlier (i.e., `ccoord` and `self._shape` should already have the correct type). The value check, if needed, I'd have earlier instead of wrapping it into a try-except.
```suggestion

            yield ZNode(
                coord=ccoord,
                shape=self._shape,
                convert_to_z=convert,
            )
```

mahdiehmalekian replied:
Agree that `TypeError` is unnecessary. However, the value check is needed, and if we want to do it earlier, it requires all the checks in `ccoord.setter` of `ZNode` to be repeated here.

mahdiehmalekian replied:
Done in a56aa8453ea0915c9c4b824e5606939d762a3561

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+            except (TypeError, ValueError):
+                pass
+
+    def external_neighbors_generator(
```
Same comments here as for the `internal_neighbors_generator` above regarding the docstring and try-except.

mahdiehmalekian replied:
Done in a56aa8453ea0915c9c4b824e5606939d762a3561

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+            except (TypeError, ValueError):
+                pass
+
+    def odd_neighbors_generator(
```
Same comments here as for the `internal_neighbors_generator` above regarding the docstring and try-except.

mahdiehmalekian replied:
Done in a56aa8453ea0915c9c4b824e5606939d762a3561

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+            raise ValueError(
+                f"Expected {self, other} to have the same shape, got {self._shape, other._shape}"
+            )
+        return self._ccoord != other._ccoord
```
This could just reuse the `__eq__` method above. You can simply remove it and let Python handle it (it now automatically negates `__eq__` if `!=` is called, so no need to implement `__ne__` at all).
```suggestion
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+        try:
+            return PlaneShift(x_shift=x_shift, y_shift=y_shift)
+        except:
+            raise ValueError(f"{other} cannot be subtracted from {self}")
```
Better be explicit.
```suggestion
        except ValueError as e:
            raise ValueError(f"{other} cannot be subtracted from {self}") from e
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,796 @@
+            raise ValueError(f"{other} cannot be subtracted from {self}")
+
+    def __hash__(self) -> int:
+        return (self._ccoord, self._shape).__hash__()
```
Not wrong per se, but it's usually better to call the hash function rather than the dunder method.
```suggestion
        return hash((self._ccoord, self._shape))
```

thisac commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,163 @@
+        return self._shift.y
+
+    def __mul__(self, scale: int | float) -> PlaneShift:
+        """Multiplies the self from left by the number value``scale``.
```
```suggestion
        """Multiplies the self from left by the number value ``scale``.
```

thisac commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,163 @@
+        new_shift_y = scale * self._shift.y
+        if int(new_shift_x) != new_shift_x or int(new_shift_y) != new_shift_y:
+            raise ValueError(f"{scale} cannot be multiplied by {self}")
+        return PlaneShift(int(new_shift_x), int(new_shift_y))
```
```suggestion
        if not isinstance(scale, (int, float)):
            raise TypeError(f"Expected scale to be int or float, got {type(scale)}")

        new_shift_x = scale * self._shift.x
        new_shift_y = scale * self._shift.y
        if int(new_shift_x) != new_shift_x or int(new_shift_y) != new_shift_y:
            raise ValueError(f"{scale} cannot be multiplied by {self}")

        return PlaneShift(int(new_shift_x), int(new_shift_y))
```

thisac commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,163 @@
+        new_shift_y = scale * self._shift.y
+        if int(new_shift_x) != new_shift_x or int(new_shift_y) != new_shift_y:
+            raise ValueError(f"{scale} cannot be multiplied by {self}")
+        return PlaneShift(int(new_shift_x), int(new_shift_y))
```
Seems like you could limit this method to only accept integers (or if you want it to work with the float type you could just check if `scale` is an integer using 
```python
if not isinstance(scale, int) or (isinstance(scale, float) and scale.is_integer()):
    raise ValueError("Scale must be an integer.")
```
replacing the other checks in the method.

mahdiehmalekian replied:
Done in 195b9cf.

thisac commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,163 @@
+        return PlaneShift(int(new_shift_x), int(new_shift_y))
+
+    def __rmul__(self, scale: int | float) -> PlaneShift:
+        """Multiplies the self from right by the number value``scale``.
```
```suggestion
        """Multiplies the self from right by the number value ``scale``.
```

thisac commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,163 @@
+        return self * scale
+
+    def __add__(self, other: PlaneShift) -> PlaneShift:
+        """
+        Adds another PlaneShift object to self.
+        Args:
```
```suggestion
        Adds another PlaneShift object to self.

        Args:
```

thisac commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,163 @@
+        return self._shift.__iter__()
+
+    def __len__(self) -> int:
+        return self._shift.__len__()
```
(minor) I'd lean towards using the available functions in these cases.
```suggestion
        return len(self._shift)
```

thisac commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,163 @@
+        return self._shift.__len__()
+
+    def __hash__(self) -> int:
+        return self._shift.__hash__()
```
```suggestion
        return hash(self._shift)
```

thisac commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,163 @@
+        return self._shift.__hash__()
+
+    def __getitem__(self, key) -> int:
+        return self._shift.__getitem__(key)
```
```suggestion
        return self._shift[key]
```

thisac commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,163 @@
+    def __ne__(self, other: PlaneShift) -> bool:
+        if not isinstance(other, PlaneShift):
+            raise TypeError(f"Expected other to be PlaneShift, got {type(other)}")
+        return not self._shift == other._shift
```
Not needed.
```suggestion
```

thisac commented on minorminer/utils/zephyr/coordinate_systems.py:
```diff
@@ -0,0 +1,71 @@
+
+def cartesian_to_zephyr(ccoord: CartesianCoord) -> ZephyrCoord:
+    """Converts a CartesianCoord to its corresponding ZephyrCoord.
+    Note: It assumes the given CartesianCoord is valid.
```
```suggestion
    """Converts a ``CartesianCoord`` to its corresponding ``ZephyrCoord``.

    Note: It assumes the given ``CartesianCoord`` is valid.
```

thisac commented on minorminer/utils/zephyr/coordinate_systems.py:
```diff
@@ -0,0 +1,71 @@
+
+def zephyr_to_cartesian(zcoord: ZephyrCoord) -> CartesianCoord:
+    """Converts a ZephyrCoord to its corresponding CartesianCoord.
+    Note: It assumes the given ZephyrCoord is valid.
```
```suggestion
    """Converts a ``ZephyrCoord``` to its corresponding ``CartesianCoord``.

    Note: It assumes the given ZephyrCoord is valid.
```

thisac commented on minorminer/utils/zephyr/survey.py:
```diff
@@ -0,0 +1,292 @@
+            G_top = G_info.get("family")
+            if G_top != "zephyr":
+                raise ValueError(f"Expected a graph with zephyr topology, got {G_top}")
+            m, t, coord = G_info.get("rows"), G_info.get("tile"), G_info.get("labels")
```
```suggestion
                raise ValueError(f"Expected a graph with zephyr topology, got {G_top}")

            m, t, coord = G_info.get("rows"), G_info.get("tile"), G_info.get("labels")
```

thisac commented on minorminer/utils/zephyr/survey.py:
```diff
@@ -0,0 +1,292 @@
+            return ZShape(m=m, t=t), coord
+
+        def sampler_shape_coord(sampler: DWaveSampler) -> dict[str, int | str]:
+            sampler_top: dict[str, str | int] = sampler.properties.get("topology")
```
Missing docstring. It's nice to have a simple one at least, for developers, even if it's not accessible or used outside of the `get_shape_coord` method.

mahdiehmalekian replied:
Done in 5ef99a2bd9751c3565b7897ae4a5280abae10a3b

thisac commented on minorminer/utils/zephyr/survey.py:
```diff
@@ -0,0 +1,292 @@
+        """
+
+        def graph_shape_coord(G: nx.Graph) -> dict[str, int | str]:
+            G_info = G.graph
```
Missing docstring.

thisac replied:
I usually like to have these internal methods be written as private methods (e.g., `_graph_shape_coord`) just to make it extra clear that these aren't used anywhere else than in this method.

mahdiehmalekian replied:
Done in 5ef99a2bd9751c3565b7897ae4a5280abae10a3b

thisac commented on minorminer/utils/zephyr/survey.py:
```diff
@@ -0,0 +1,292 @@
+        if isinstance(G, DWaveSampler):
+            return sampler_shape_coord(G)
+        else:
+            raise TypeError(f"Expected G to be networkx.Graph or DWaveSampler, got {type(G)}")
```
```suggestion

        raise TypeError(f"Expected G to be networkx.Graph or DWaveSampler, got {type(G)}")
```

thisac commented on minorminer/utils/zephyr/survey.py:
```diff
@@ -0,0 +1,292 @@
+
+    @property
+    def shape(self) -> ZShape:
+        """Returns the ZShape of G"""
```
```suggestion
        """Returns the ``ZShape of ``G``."""
```

mahdiehmalekian replied:
Done in 5ef99a2bd9751c3565b7897ae4a5280abae10a3b

thisac commented on minorminer/utils/zephyr/survey.py:
```diff
@@ -0,0 +1,292 @@
+
+    @property
+    def nodes(self) -> set[ZNode]:
+        """Returns the ZNodes of the sampler/graph"""
```
```suggestion
        """Returns the ``ZNode``s of the sampler/graph."""
```

thisac commented on minorminer/utils/zephyr/survey.py:
```diff
@@ -0,0 +1,292 @@
+        if not isinstance(v, ZNode):
+            raise TypeError(f"Expected v to be ZNode, got {v}")
+        if v in self.missing_nodes:
+            return {}
```
I assume this should return an empty set (`{}` creates an empty dict).
```suggestion
            return set()
```

mahdiehmalekian replied:
Done in 5ef99a2bd9751c3565b7897ae4a5280abae10a3b

thisac commented on minorminer/utils/zephyr/survey.py:
```diff
@@ -0,0 +1,292 @@
+
+    def _ext_path(self, uwkj: UWKJ) -> set[ZSE]:
+        """
+        Returns uwkj_sur, where uwkj_sur contains ZSE(z_start, z_end)
```
```suggestion
        """Returns uwkj_sur, where uwkj_sur contains ZSE(z_start, z_end)
```

thisac commented on minorminer/utils/zephyr/survey.py:
```diff
@@ -0,0 +1,292 @@
+        """Returns the degree of v when restricted to nbr_kind"""
+        return len(self.neighbors(v, nbr_kind=nbr_kind))
+
+    def _ext_path(self, uwkj: UWKJ) -> set[ZSE]:
```
Type hinted as returning a set, but actually returns a list.

mahdiehmalekian replied:
Corrected in 5ef99a2bd9751c3565b7897ae4a5280abae10a3b

thisac commented on minorminer/utils/zephyr/survey.py:
```diff
@@ -0,0 +1,292 @@
+
+        def _ext_seg(z_start: int) -> ZSE | None:
+            """
+            If (u, w, k, j, z_start) does not exist, returns None.
```
```suggestion
            """If (u, w, k, j, z_start) does not exist, returns None.
```

thisac commented on minorminer/utils/zephyr/survey.py:
```diff
@@ -0,0 +1,292 @@
+            is_extensible = _ext_seg(z_start + 1)
+            if is_extensible is None:
+                return ZSE(z_start, z_start)
+            return ZSE(z_start, is_extensible.z_end)
```
```suggestion
            cur: ZNode = ZNode(coord=ZephyrCoord(*uwkj, z_start), shape=self.shape)
            if cur in self.missing_nodes:
                return None

            upper_z: int = z_vals[-1]
            if z_start > upper_z:
                return None

            if z_start == upper_z:
                return ZSE(z_start, z_start)

            next_ext = ZNode(coord=ZephyrCoord(*uwkj, z_start + 1), shape=self.shape)
            ext_edge = ZEdge(cur, next_ext)
            if ext_edge in self.missing_edges:
                return ZSE(z_start, z_start)

            is_extensible = _ext_seg(z_start + 1)
            if is_extensible is None:
                return ZSE(z_start, z_start)

            return ZSE(z_start, is_extensible.z_end)
```

thisac commented on minorminer/utils/zephyr/survey.py:
```diff
@@ -0,0 +1,292 @@
+    def calculate_external_paths_stretch(self) -> dict[UWKJ, set[ZSE]]:
+        """
+        Returns {uwkj: set of connected segments (z_start, z_end)}
+        """
```
Incomplete docstring (no description, only returns.

mahdiehmalekian replied:
Done in 5ef99a2bd9751c3565b7897ae4a5280abae10a3b

thisac commented on minorminer/utils/zephyr/qfloor.py:
```diff
@@ -0,0 +1,402 @@
+
+
+class QuoTile:
+    """Initializes a 'QuoTile' object with zns.
```
Better to be explicit.
```suggestion
    """Initializes a 'QuoTile' object with ``ZNode``.
```

mahdiehmalekian replied:
Done in 659c815.

thisac commented on minorminer/utils/zephyr/qfloor.py:
```diff
@@ -0,0 +1,402 @@
+
+    @cached_property
+    def index_zns(self) -> dict[int, ZNode]:
+        return {i: zn for i, zn in enumerate(self._zns)}
```
Missing docstrings for a lot of these methods.

mahdiehmalekian replied:
Done in 241667a401728fce80a3a6f0f4c047567a7d9084

thisac commented on minorminer/utils/zephyr/qfloor.py:
```diff
@@ -0,0 +1,402 @@
+
+    @property
+    def seed(self) -> ZNode:
+        """Returns the smallest ZNode in Zns"""
```
(minor) Missing punctuation after line in a few of these methods.
```suggestion
        """Returns the smallest ZNode in Zns."""
```

thisac commented on minorminer/utils/zephyr/qfloor.py:
```diff
@@ -0,0 +1,402 @@
+    >>> zns = [ZNode(coord=c) for c in coords]
+    >>> floor = QuoFloor(corner_qtile=zns, dim=(3, 5))
+    >>> print(f"{floor.qtile_xy(2, 3) = }")
+    floor.qtile_xy(2, 3) = QuoTile[ZNode(CartesianCoord(x=8, y=13, k=None)), ZNode(CartesianCoord(x=9, y=12, k=None)), ZNode(CartesianCoord(x=9, y=14, k=None)), ZNode(CartesianCoord(x=10, y=13, k=None)), ZNode(CartesianCoord(x=10, y=15, k=None)), ZNode(CartesianCoord(x=11, y=14, k=None)), ZNode(CartesianCoord(x=11, y=16, k=None)), ZNode(CartesianCoord(x=12, y=15, k=None))]
```
```suggestion

    Example:
    .. code-block:: python

        >>> from minorminer.utils.zephyr.node_edge import ZNode
        >>> from minorminer.utils.zephyr.qfloor import QuoFloor
        >>> coords = [(k, k+1) for k in range(4)] + [(k+1, k) for k in range(4)]
        >>> zns = [ZNode(coord=c) for c in coords]
        >>> floor = QuoFloor(corner_qtile=zns, dim=(3, 5))
        >>> print(f"{floor.qtile_xy(2, 3) = }")
        floor.qtile_xy(2, 3) = QuoTile[ZNode(CartesianCoord(x=8, y=13, k=None)), ZNode(CartesianCoord(x=9, y=12, k=None)), ZNode(CartesianCoord(x=9, y=14, k=None)), ZNode(CartesianCoord(x=10, y=13, k=None)), ZNode(CartesianCoord(x=10, y=15, k=None)), ZNode(CartesianCoord(x=11, y=14, k=None)), ZNode(CartesianCoord(x=11, y=16, k=None)), ZNode(CartesianCoord(x=12, y=15, k=None))]
```

thisac commented on minorminer/utils/zephyr/qfloor.py:
```diff
@@ -0,0 +1,402 @@
+    implemented_connectors: tuple[dict[tuple[int], PlaneShift]] = (
+        {(1, 0): PlaneShift(4, 0), (0, 1): PlaneShift(0, 4)},
+    )
+    tile_connector0 = implemented_connectors[0]
```
Maybe capitalize these since they're constants?

mahdiehmalekian replied:
Done in 66e65ba

thisac reviewed (commented):
Had another review round. Haven't checked the tests yet (will do that next). 

A couple of overall comments and suggestions:

* `qfloor.py` contains a couple of methods without or with incomplete docstrings.
* Code in docstrings should have double back-ticks (``` `` ```), and wherever possible, it's nice (but not mandatory) to have cross-references to classes and methods, etc. See [here](https://www.sphinx-doc.org/en/master/usage/referencing.html).
* A couple of places I'd recommend adding a new line after returns and around for-loop-blocks, try-excepts, after raises and after returns, to make it a bit more legible.

jackraymond committed:
implement review changes to docstrings

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/plane_shift.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/plane_shift.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/plane_shift.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/plane_shift.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/plane_shift.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/plane_shift.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/plane_shift.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/plane_shift.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/survey.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/survey.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/survey.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/survey.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/survey.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/survey.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/coordinate_systems.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/coordinate_systems.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/qfloor.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/qfloor.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

Commit:
capitalize

Commit:
Update docstring

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

Commit:
Disable 'float' scale

Commit:
Correct header

Commit:
Add, correct documentation

Commit:
Add, correct documentation

Commit:
Correct documentation

Commit:
Correct documentation

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

Commit:
Correct documentation

Commit:
Disable shape.setter, coord.setter + Separate checks for ccoord + Modify internal, external, odd neighbors generators+ Edit the docstrings

Commit:
Edit docstrings

jackraymond committed:
Allow dwave_networkx to be passed as kwargs rather than in T

jackraymond committed:
Bugfix: remove None test for required List argument

jackraymond committed:
Apply suggestions from code review

Co-authored-by: Radomir Stevanovic <radomir.stevanovic@gmail.com>

jackraymond committed:
Fix Optional type hinting, fix handling of T=None code and tests

randomir committed:
Release 0.2.19

### New Features

- Allow lattice type and lattice dimensions to be passed as options for `find_sublattice_embeddings()`. See [\#266](https://github.com/dwavesystems/minorminer/pull/266).

### Bug Fixes

- Fix `find_sublattice_embeddings()` to not produce non-disjoint embeddings when `use_tile_embedding=True` is set. See [\#265](https://github.com/dwavesystems/minorminer/pull/265).

boothby commented:
Thank you for this contribution!  Let's meet up and chat about what parts of this can be generalized.  We still have Advantage systems in productions and it would be nice to see if we can extend this to work with Pegasus graphs.  And even if not, the Chimera structures underlying both Pegasus and Zephyr can be leveraged to support Pegasus if at lesser embedding efficiencies.

I wouldn't burden this PR with such a high bar, but I think that it would be worth thinking through what the user interface would look like (e.g. is `minorminer.utils.zephyr` the right namespace?) to make sure that we can cleanly support other architectures.

mahdiehmalekian commented:
@boothby  Most of the tools in this PR are quite specific to the zephyr architecture, although similar tools can definitely be developed for future architectures and/or previous ones. Happy to chat about this and also the user interface. Let's have @thisac also in the loop.

mahdiehmalekian commented:
@randomir is this ready to be merged?

randomir commented:
> @randomir is this ready to be merged?

Hey @mahdiehmalekian, let me do a quick review first please (I didn't want to comment it while you were away). Also, we need to fix tests before we can merge. I can have a look why tests are failing (unless you already know why).

mahdiehmalekian commented:
@randomir Thank you.
Re. test failing, I believe @thisac said it has to do with `dwave.system.DWaveSampler` or `dwave.cloud.Client` being used in the test suite and that he'd look into how to work around it. I couldn't solve it myself. Would appreciate if you can help with it.

boothby commented:
I'd like to have that conversation about how we arrange topology-targeting tools within the package before merging, because maintaining backwards-compatibility turns into a maintenance chore we can otherwise avoid.  For an immediate example, PlaneShift is not specific to the Zephyr topology, and we'd not want to duplicate that for a Chimera- or Pegasus-specific toolset.  Leaving that in the zephyr namespace would push undue burden of refactoring and deprecating onto future developers.

boothby committed:
Added support for colorings in Glasgow subgraph solver (#272)

* Added support for colorings in Glasgow subgraph solver
* updated Glasgow to latest version (with patches for c++17 support)
* new parameters for subgraph.find_subgraph
  * as_embedding=True will produce a dict with tuple values
  * node_labels to be respected by GSS, node labels must match
  * edge_labels as with node_labels; directed edge labels match

boothby committed:
Glasgow Miscellany (#273)

* removed kwargs handling in favor for explicit arguments (issue #255)
* added injectivity parameter to support noninjective and locally-injective homomorphisms
* fix isolated nodes issue #254 and add injectivity options
* support random seeds (issue #243)

boothby commented:
@mahdiehmalekian it looks to me like the only place that you're using the cloud client is to load up some zephyr graphs with defects.  My standard practice for this is to (a) generate a list of defects randomly during the testing process, or (b) collect a list from a real-world example, and hard-code that list into the test.

thisac committed:
Add Python 3.14 and remove 3.9 support

thisac committed:
Set default start method for multiprocessing to 'fork'

jackraymond commented on minorminer/utils/zephyr/coordinate_systems.py:
```diff
@@ -0,0 +1,73 @@
+    if x % 2 == 0:
+        u: int = 0
+        w: int = x // 2
+        j: int = ((y - 1) % 4) // 2
```
t=4 is hard coded here, can it be made to work for any t, w.r.t. Zephyr[m,t] graphs?

I'm interested in the relation between this coordinate system and https://github.com/dwavesystems/dwave-hybrid/blob/master/hybrid/decomposers.py#L864 as well

jackraymond reviewed (commented):
The cartesian system looks a bit like that used here: 
https://github.com/dwavesystems/dwave-hybrid/blob/master/hybrid/decomposers.py#L864
but this scheme has a lot more power and might be useful to simplify the hybrid codes - ideally it could work with zephyr_torus, but not essential.  There are also cartesian displacements hidden in things like zephyr_sublattice_mappings that might be tidied up?

I assume once this goes through the dependent work done by @mahdiehmalekian on cubic/biclique/clique can move forward. Some of these graph abstractions like QuoTile are useful in that context I'm assuming.

Some of the transformations have the tile parameter 't' of Zephyr[m,t] graphs hard coded, would be great to make flexible with respect to t at t=4 (I use t<4 a lot, and processor direction could move towards support of higher t in principle). 
 
I guess there is not sufficient graphical similarity between Zephyr and Pegasus/Chimera to allow some of these functions to be generalized? Happy with a specialist code.

I'm just commenting, as I haven't done a deep dive on tests/etc.

thisac committed:
Bump release to 0.2.20

thisac committed:
Update releasenote entry to fix reno parsing

boothby committed:
Add keyboard interrupts for subgraph.find_subgraph (#275)

* added ability to interrupt find_subgraph via ctrl-c

thisac committed:
Increment version for release

Commit:
Fix bug

Commit:
Add check node validity flag

Commit:
Edit examples

Commit:
Remove unused files

Commit:
Refactor, Add documentation

mahdiehmalekian commented:
@boothby
`PlaneShift` now works for a general displacement in a plane with a child class `ZPlaneShift` for Zephyr.

mahdiehmalekian commented:
@jackraymond 
Thanks for the review and the feedback.

The relation with the chimera like coordinate in [decomposers.py](https://github.com/dwavesystems/dwave-hybrid/blob/master/hybrid/decomposers.py#L864) is (r, c, u, k) -> (2c + u, 2r +1 - u, k)

The 4 in [coordinate_systems.py](https://github.com/dwavesystems/minorminer/blob/8d6c9743daaa52f0a6c39d2d1bb64ef6c45dbd84/minorminer/utils/zephyr/coordinate_systems.py#L46) doesn't refer to tile size. The coordinate system and all the tools indeed work for Zephyr with any tile size.

Unfortunately not much of this can be used for Pegasus. Although similar tools can surely be developed.

mahdiehmalekian commented:
@boothby Thanks for the suggestion re. getting around Client issue. Done, but still tests failing. @thisac said he'd have a look.

thisac commented on minorminer/utils/zephyr/coordinate_systems.py:
```diff
@@ -0,0 +1,73 @@
+
+from __future__ import annotations
+
+from collections import namedtuple
+
```
Would add an `__all__` to this file as well as `node_edge.py` and `plane_shift.py`.
```python
__all__ = ["cartesian_to_zephyr", "zephyr_to_cartesian"]
```
Not sure if `zephyr_fields`, `ZephyrCoord`, etc. below should be there as well.

mahdiehmalekian replied:
Done in [d8c9790](https://github.com/dwavesystems/minorminer/pull/261/commits/d8c979050485936ff4bf44293dbc00e1b938a61f)

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,819 @@
+    HORIZONTAL = 1
+
+
+
```
```suggestion


```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,819 @@
+
+    Args:
+        x : One endpoint of edge.
+        y : Another endpoint of edge.
```
```suggestion
        x: One endpoint of edge.
        y: Another endpoint of edge.
```

mahdiehmalekian replied:
Done in [d8c9790](https://github.com/dwavesystems/minorminer/pull/261/commits/d8c979050485936ff4bf44293dbc00e1b938a61f)

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,819 @@
+        self,
+        x,
+        y,
+    ) -> None:
```
Should likely be on a single line, and also missing typehints.

mahdiehmalekian replied:
Done in [d8c9790](https://github.com/dwavesystems/minorminer/pull/261/commits/d8c979050485936ff4bf44293dbc00e1b938a61f)

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,819 @@
+
+
+class Edge:
+    """Initializes an Edge with nodes x, y.
```
This docstring could be slightly expanded. Also, the class doesn't per se initialize an edge, but rather it _represents_ an edge, so slight change.

mahdiehmalekian replied:
Done in [d8c9790](https://github.com/dwavesystems/minorminer/pull/261/commits/d8c979050485936ff4bf44293dbc00e1b938a61f)

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,819 @@
+    ) -> None:
+        self._edge = self._set_edge(x, y)
+
+    def _set_edge(self, x, y):
```
Missing typehints.

mahdiehmalekian replied:
Done in [d8c9790](https://github.com/dwavesystems/minorminer/pull/261/commits/d8c979050485936ff4bf44293dbc00e1b938a61f)

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,819 @@
+        self._edge = self._set_edge(x, y)
+
+    def _set_edge(self, x, y):
+        """Reutrns ordered tuple corresponding to the set {x, y}."""
```
```suggestion
        """Returns ordered tuple corresponding to the set {x, y}."""
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,819 @@
+    def __getitem__(self, index: int) -> int:
+        return self._edge[index]
+
+    def __eq__(self, other: Edge):
```
```suggestion
    def __eq__(self, other: Edge) -> bool:
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,819 @@
+
+    Args:
+        x (ZNode): One endpoint of edge.
+        y (ZNode): Another endpoint of edge.
```
A few of my older comments seem to have been missed. Readding them.
```suggestion
        x (ZNode): Endpoint of edge. Must have same shape as ``y``.
        y (ZNode): Endpoint of edge. Must have same shape as ``x``
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,819 @@
+        m: grid size, t: tile size
+        Defaults to None.
+        convert_to_z (bool | None, optional): Whether to express the coordinates in ZephyrCoordinates.
+        Defaults to None.
```
```suggestion
        convert_to_z (bool | None, optional): Whether to express the coordinates in ZephyrCoordinates.
            Defaults to None.
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,819 @@
+    """Initializes 'ZNode' with coord and optional shape.
+
+    Args:
+        coord (CartesianCoord | ZephyrCoord | tuple[int]): coordinate in (quotient) Zephyr or (quotient) Cartesian
```
```suggestion
        coord (CartesianCoord | ZephyrCoord | tuple[int]): Coordinate in (quotient) Zephyr or (quotient) Cartesian.
```

thisac commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,819 @@
+
+    Note: If the given coord has non-None k value (in either Cartesian or Zephyr coordinates),
+        shape = None raises ValueError. In this case the tile size of Zephyr, t,
+        must be provided.
```
```suggestion
    ..note::
    
        If the given coord has non-None k value (in either Cartesian or Zephyr coordinates),
        ``shape = None`` raises ValueError. In this case the tile size of Zephyr, t,
        must be provided.
```

thisac commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,137 @@
+        return type(self) == type(other) and self._xy == other._xy
+
+
+
```
```suggestion


```

thisac commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,137 @@
+
+
+class ZPlaneShift(PlaneShift):
+    """Initializes ZPlaneShift with an x, y.
```
Expand on this docstring to describe that, unlike `PlaneShift` above this is Zephyr specific.

mahdiehmalekian replied:
Done in [d8c9790](https://github.com/dwavesystems/minorminer/pull/261/commits/d8c979050485936ff4bf44293dbc00e1b938a61f)

thisac commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,137 @@
+        self,
+        x: int,
+        y: int,
+    ) -> None:
```
Would be cleaner on a single line.

mahdiehmalekian replied:
Done in [d8c9790](https://github.com/dwavesystems/minorminer/pull/261/commits/d8c979050485936ff4bf44293dbc00e1b938a61f)

thisac commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,137 @@
+        if x % 2 != y % 2:
+            raise ValueError(
+                f"Expected x, y to have the same parity, got {x, y}"
+            )
```
Are these checks what make this Zephyr specific?

mahdiehmalekian replied:
Yes

thisac commented on tests/utils/zephyr/test_coordinate_systems.py:
```diff
@@ -0,0 +1,62 @@
+# Copyright 2025 D-Wave Systems Inc.
```
```suggestion
# Copyright 2025 D-Wave
```

thisac commented on tests/utils/zephyr/test_node_edge.py:
```diff
@@ -0,0 +1,366 @@
+# Copyright 2025 D-Wave Systems Inc.
```
```suggestion
# Copyright 2025 D-Wave
```

thisac commented on tests/utils/zephyr/test_node_edge.py:
```diff
@@ -0,0 +1,366 @@
+                                for k in k_vals:
+                                    ZNode(coord=(u, w, k, j, z), shape=(m, t))
+
+    def test_zephyr_node_invalid_args_raises_error(self) -> None:
```
Could all of the asserts below be separated into subtests or be parameterized? I think that would be better to make this test a bit cleaner.

mahdiehmalekian replied:
Done in [d8c9790](https://github.com/dwavesystems/minorminer/pull/261/commits/d8c979050485936ff4bf44293dbc00e1b938a61f)

thisac commented on tests/utils/zephyr/test_node_edge.py:
```diff
@@ -0,0 +1,366 @@
+
+    def test_neighbors_generator_runs(self) -> None:
+        for _ in ZNode((1, 12, 4), ZShape(t=6)).neighbors_generator():
+            _
```
This test seems a bit lacking. Perhaps test that the neighbours are correct as well? Also, better to use `pass` instead of `_` here if nothing happens.

mahdiehmalekian replied:
Done in [d8c9790](https://github.com/dwavesystems/minorminer/pull/261/commits/d8c979050485936ff4bf44293dbc00e1b938a61f)

thisac commented on tests/utils/zephyr/test_node_edge.py:
```diff
@@ -0,0 +1,366 @@
+        self.assertTrue(zn.neighbor_kind(ZNode((0, 3))) is EdgeKind.ODD)
+        self.assertTrue(zn.neighbor_kind(ZNode((0, 5))) is EdgeKind.EXTERNAL)
+        self.assertTrue(zn.neighbor_kind(ZNode((0, 7))) is None)
+        self.assertTrue(zn.neighbor_kind(ZNode((1, 6))) is None)
```
This could definitely be [parameterized](https://github.com/wolever/parameterized) instead, since they're covering different cases.

randomir replied:
Also, you can use a specialized assert method, [`assertIs`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertIs), for cleaner error messages.

mahdiehmalekian replied:
Done in [d8c9790](https://github.com/dwavesystems/minorminer/pull/261/commits/d8c979050485936ff4bf44293dbc00e1b938a61f)

thisac commented on tests/utils/zephyr/test_node_edge.py:
```diff
@@ -0,0 +1,366 @@
+        expected2 = {ZNode((1, 0, k), ZShape(t=4)) for k in range(4)} | {
+            ZNode((1, 2, k), ZShape(t=4)) for k in range(4)
+        }
+        self.assertEqual(set_internal2, expected2)
```
Should probably be separate tests, or parameterized also. Same for the ones below.

mahdiehmalekian replied:
Done in [d8c9790](https://github.com/dwavesystems/minorminer/pull/261/commits/d8c979050485936ff4bf44293dbc00e1b938a61f)

thisac commented on tests/utils/zephyr/test_node_edge.py:
```diff
@@ -0,0 +1,366 @@
+        expected2 = {ZNode((0, 3, 2), ZShape(t=4))}
+        self.assertEqual(set_odd2, expected2)
+
+    def test_degree(self) -> None:
```
Same for these, with regards to using [parameterized](https://github.com/wolever/parameterized) or [subtests](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests).

mahdiehmalekian replied:
Done in [d8c9790](https://github.com/dwavesystems/minorminer/pull/261/commits/d8c979050485936ff4bf44293dbc00e1b938a61f)

thisac commented on tests/utils/zephyr/test_plane_shift.py:
```diff
@@ -0,0 +1,95 @@
+# Copyright 2025 D-Wave Systems Inc.
```
```suggestion
# Copyright 2025 D-Wave
```

thisac reviewed (commented):
The issue with the tests is related to not finding the utils.zephyr subpackage. Easiest way to fix should be to add `'minorminer.utils.zephyr'` to `setup.py`, under packages:
```
setup(
    packages=[
        ...
        'minorminer.utils.zephyr',
        ...
    ],
)
```

randomir commented:
Once the above comments are addressed, @mahdiehmalekian, please rebase your branch to the upstream `main` - to make sure we run tests on python 3.10 through 3.14.

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update tests/utils/zephyr/test_plane_shift.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/plane_shift.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update minorminer/utils/zephyr/node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update tests/utils/zephyr/test_coordinate_systems.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

mahdiehmalekian committed:
Update tests/utils/zephyr/test_node_edge.py

Co-authored-by: Theodor Isacsson <theodor@isacsson.ca>

Commit:
Include utils.zephyr in packages

Commit:
Apply suggestions from code review

Commit:
Merge remote-tracking branch 'upstream' into zephyr-utils

randomir commented on tests/utils/zephyr/test_coordinate_systems.py:
```diff
@@ -0,0 +1,68 @@
+        xyks = [(0, 1), (1, 0, None), (12, -3)]
+        ccoords = [CartesianCoord(*xyk) for xyk in xyks]
+        for ccoord in ccoords:
+            cartesian_to_zephyr(ccoord=ccoord)
```
I would fold the two `*_runs` tests with the parameterized ones, but if you want to keep them as smoke tests, I recommend at least testing the returned type:

```suggestion
            self.assertIsInstance(cartesian_to_zephyr(ccoord=ccoord), ZephyrCoord)
```

randomir commented on tests/utils/zephyr/test_coordinate_systems.py:
```diff
@@ -0,0 +1,68 @@
+        uwkjzs = [(0, 2, 4, 1, 5), (1, 3, 3, 0, 0), (1, 2, None, 1, 5)]
+        zcoords = [ZephyrCoord(*uwkjz) for uwkjz in uwkjzs]
+        for zcoord in zcoords:
+            zephyr_to_cartesian(zcoord=zcoord)
```
Same comment as above:

```suggestion
            self.assertIsInstance(zephyr_to_cartesian(zcoord=zcoord), CartesianCoord)
```

randomir commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,127 @@
+    """Represents a displacement in a Cartesian plane.
+
+    Args:
+        x (int): The displacement in the x-direction of a Cartesian coordinate.
+        y (int): The displacement in the y-direction of a Cartesian coordinate.
```
In Ocean, we follow [Google Python Style Guide](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods), and omit types from docstrings when they are type-annotated. Here, and elsewhere in this PR.

randomir commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,127 @@
+        Returns:
+            PlaneShift: The displacement in CartesianCoord by self followed by other.
+        """
+        if type(self) != type(other):
```
By using `isinstance`, you support `PlaneShift` subclasses for `other`:

```suggestion
        if not isinstance(other, PlaneShift):
```

randomir commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,127 @@
+
+from __future__ import annotations
+
+from typing import Iterator
```
[Deprecated](https://docs.python.org/3/library/typing.html#typing.Iterator) in favor of:

```suggestion
from collections.abc import Iterator
```

randomir commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,127 @@
+        return type(self)(self.x + other.x, self.y + other.y)
+
+    def __iter__(self) -> Iterator[int]:
+        return self._xy.__iter__()
```
Like for `len` and `hash`, you can use `iter()`:

```suggestion
        return iter(self._xy)
```

randomir commented on minorminer/utils/zephyr/plane_shift.py:
```diff
@@ -0,0 +1,127 @@
+            raise ValueError(
+                f"Expected x, y to have the same parity, got {x, y}"
+            )
+        self._xy = (x, y)
```
To ensure the superclass gets fully initialized (imagine in the future `PlaneShift` init does something in addition to `_xy` assignment):

```suggestion
        super().__init__(x, y)
```

randomir commented on tests/utils/zephyr/test_plane_shift.py:
```diff
@@ -0,0 +1,105 @@
+            (0, 0),
+        ]
+
+    def test_valid_input_runs(self) -> None:
+        for shift in self.shifts:
+            PlaneShift(*shift)
```
Looks like you're testing construction here:

```suggestion
    def test_construction(self) -> None:
        for shift in self.shifts:
            p = PlaneShift(*shift)
            self.assertEqual(shift, (p.x, p.y))
```

randomir commented on tests/utils/zephyr/test_plane_shift.py:
```diff
@@ -0,0 +1,105 @@
+            (0, 0),
+        ]
+
+    def test_valid_input_runs(self) -> None:
+        for shift in self.shifts:
+            ZPlaneShift(*shift)
```
Again, it's best to assert the expected result, not settle with "not failing":

```suggestion
    def test_construction(self) -> None:
        for shift in self.shifts:
            zp = ZPlaneShift(*shift)
            self.assertEqual(shift, (zp.x, zp.y))
```

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+        return self._edge == other._edge
+
+    def __str__(self) -> str:
+        return f"{type(self).__name__}{self._edge}"
```
Simpler format worth considering:

```suggestion
        return str(self._edge)
```

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+            else:
+                raise ValueError(f"Expected x, y to be neighbors, got {x, y}")
+
+        self._edge = self._set_edge(x, y)
```
Use `super()`'s init to ensure proper initialization and maintain abstraction:

```suggestion
        super().__init__(x, y)
```

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+        y: ZNode,
+        check_edge_valid: bool = True,
+    ) -> None:
+        if check_edge_valid:
+            if not isinstance(x, ZNode) or not isinstance(y, ZNode):
+                raise TypeError(f"Expected x, y to be ZNode, got {type(x), type(y)}")
+
+            if x.shape != y.shape:
+                raise ValueError(f"Expected x, y to have the same shape, got {x.shape, y.shape}")
+
```
Type and shape are basic checks that should always be performed, IMO:

```suggestion
        if not isinstance(x, ZNode) or not isinstance(y, ZNode):
            raise TypeError(f"Expected x, y to be ZNode, got {type(x), type(y)}")

        if x.shape != y.shape:
            raise ValueError(f"Expected x, y to have the same shape, got {x.shape, y.shape}")

        if check_edge_valid:
```

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+            if x.shape != y.shape:
+                raise ValueError(f"Expected x, y to have the same shape, got {x.shape, y.shape}")
+
+            for kind in EdgeKind:
+                if x.is_neighbor(y, nbr_kind=kind):
+                    self._edge_kind = kind
+                    break
+            else:
+                raise ValueError(f"Expected x, y to be neighbors, got {x, y}")
```
Since you already do practically the same check in `edge_kind` property, you could simplify this to:

```suggestion
            # edge_kind property calculation verifies edge is a valid Zephyr edge
            self.edge_kind
```

If you apply this, you'll also need to move the init above.

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+    """Represents a node of a graph with Zephyr topology with coordinate and optional shape.
+
+    Args:
+        coord (CartesianCoord | ZephyrCoord | tuple[int]): Coordinate in (quotient) Zephyr or (quotient) Cartesian.
+        shape (ZShape | tuple[int | None] | None, optional): shape of Zephyr graph containing ZNode.
```
As noted in another comment, if type annotation is used, no need to replicate types in docstrings.

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+    Args:
+        coord (CartesianCoord | ZephyrCoord | tuple[int]): Coordinate in (quotient) Zephyr or (quotient) Cartesian.
+        shape (ZShape | tuple[int | None] | None, optional): shape of Zephyr graph containing ZNode.
+        m: grid size, t: tile size
+        Defaults to None.
```
misaligned

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+    """
+    def __init__(
+        self,
+        coord: CartesianCoord | ZephyrCoord | tuple[int],
```
Invalid [`tuple` annotation](https://docs.python.org/3/library/typing.html#annotating-tuples).

I believe you want:
```python
tuple[int, ...]
```

Or if you want to limit it 2-, 3-, 4-, and 5-tuples (not very elegant, but at least can create as alias):
```
type CoordTuple = tuple[int, int] | tuple[int, int, int] | tuple[int, int, int, int] | tuple[int, int, int, int, int]
```

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+        m: grid size, t: tile size
+        Defaults to None.
+        convert_to_z (bool | None, optional): Whether to express the coordinates in ZephyrCoordinates.
+            Defaults to None.
```
We should clarify what `None` means -- convert depends on `coords` type.

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+
+    @property
+    def node_kind(self) -> NodeKind:
+        """Returns the :class:`NodeKind` of ``self``."""
```
Two guidelines that apply not just to this property:
- [property docstring](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) should be consistent with an attribute docstring
- refer to type, instead "self"
```suggestion
        """The :class:`NodeKind` of the ``ZNode``."""
```

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+        if y1 != y2:  # odd, external neighbors only on the same horizontal lines
+            return None
+
+        diff_x = abs(x1 - x2)
```
Could move `diff_x` and `diff_y` up to reuse it.

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+        """Returns True if the node represents a horizontal qubit (i.e., its ``u`` coordinate in Zephyr coordinates is 1)."""
+        return self.node_kind is NodeKind.HORIZONTAL
+
+    def neighbor_kind(
```
Would `neighbor_edge_kind` fit better here?

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+        """Generator of internal neighbors of self when restricted by ``where``.
+        Args:
+            where (Callable[[CartesianCoord | ZephyrCoord], bool]):
+                A coordinate filter. Applies to ``ccoord`` if :py:attr:`self.convert_to_z` is ``False``,
```
```suggestion
                A coordinate filter. Applies to ``ccoord`` if :attr:`ZNode.convert_to_z` is ``False``,
```

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+        else:
+            kinds = set(nbr_kind)
+        if EdgeKind.INTERNAL in kinds:
+            for cc in self.internal_neighbors_generator(where=where):
+                yield cc
```
Perfect use case for `yield from` (here, and below):

```suggestion
            yield from self.internal_neighbors_generator(where=where)
```

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+            for cc in self.odd_neighbors_generator(where=where):
+                yield cc
+
+    def neighbors(
```
What's the reason to include both the generators for all kinds of neighbors, and then redundant functions that return a set?

I would remove `_generator` suffix from the generators, and keep only those.

mahdiehmalekian replied:
Sounds good. Thanks for the suggestion.

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+        Returns:
+            bool: Whether other is an internal neighbor of self when restricted by ``where``.
+        """
+        for nbr in self.internal_neighbors_generator(where=where):
+            if other == nbr:
+                return True
+        return False
```
This, and the similar 3 other functions feel redundant as they can be replaced with:

```suggestion
        return other in self.internal_neighbors_generator(where=where)
```

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+        self,
+        nbr_kind: EdgeKind | Iterable[EdgeKind] | None = None,
+        where: Callable[[CartesianCoord | ZephyrCoord], bool] = lambda coord: True,
+    ) -> list[ZEdge]:
```
Why not make this function a generator?

randomir commented on minorminer/utils/zephyr/node_edge.py:
```diff
@@ -0,0 +1,823 @@
+        return self._ccoord == other._ccoord
+
+
+    def __gt__(self, other: ZNode) -> bool:
```
Instead of manually defining all comparison methods, use [`@functools.total_ordering`](https://docs.python.org/3/library/functools.html#functools.total_ordering).

randomir commented on tests/utils/zephyr/test_node_edge.py:
```diff
@@ -0,0 +1,350 @@
+        ]
+    )
+    def test_degree(self, xy, m, nbr_kind, a, b) -> None:
+            for t in [None, 1, 4, 6]:
```
check indent

randomir reviewed (commented):
Very well written code and super useful utils! Thank you, @mahdiehmalekian!

My feedback is mostly about code/docs style (for consistency with Ocean) and coding best practices.

In terms of interface/API, I would simplify `ZNode` by defining neighbor generators only, also removing membership test functions.
